kotlin-result-coroutines/src/jsTest/kotlin/com.github.michaelbull.result.coroutines/binding/AsyncSuspendableBindingTests.kt

```diff
@@ -0,0 +1,74 @@
+package com.github.michaelbull.result.coroutines.binding
+
+import com.github.michaelbull.result.Err
+import com.github.michaelbull.result.Ok
+import com.github.michaelbull.result.Result
+import com.github.michaelbull.result.coroutines.runBlockingTest
+import kotlinx.coroutines.async
+import kotlinx.coroutines.delay
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertTrue
+
+class AsyncSuspendableBindingTest {
+
+    private sealed class BindingError {
+        object BindingErrorA : BindingError()
+        object BindingErrorB : BindingError()
+    }
+
+    @Test
+    fun returnsOkIfAllBindsSuccessful(): dynamic {
+        suspend fun provideX(): Result<Int, BindingError> {
+            delay(100)
+            return Ok(1)
+        }
+
+        suspend fun provideY(): Result<Int, BindingError> {
+            delay(100)
+            return Ok(2)
+        }
+
+        return runBlockingTest {
+            val result = binding<Int, BindingError> {
+                val x = async { provideX().bind() }
+                val y = async { provideY().bind() }
+                x.await() + y.await()
+            }
+            assertTrue(result is Ok)
+            assertEquals(
+                expected = 3,
+                actual = result.value
+            )
+        }
+    }
+
+    @Test
+    fun returnsAnyErrIfBindingFailed() {
+        suspend fun provideX(): Result<Int, BindingError> {
+            delay(1)
+            return Ok(1)
+        }
+
+        suspend fun provideY(): Result<Int, BindingError.BindingErrorA> {
+            delay(2)
+            return Err(BindingError.BindingErrorA)
+        }
+
+        suspend fun provideZ(): Result<Int, BindingError.BindingErrorB> {
+            delay(1)
+            return Err(BindingError.BindingErrorB)
+        }
+
+        runBlockingTest {
+            val result = binding<Int, BindingError> {
+                val x = async { provideX().bind() }
+                val y = async { provideY().bind() }
+                val z = async { provideZ().bind() }
+                x.await() + y.await() + z.await()
+            }
+
+            assertTrue(result is Err)
```

Comment:
thanks for adding 😎  
Now I'm curious, is there just no guarantee which we will receive first in this case on JS? or will it always be `BindingErrorA` due to the order the `await`s are invoked? If so, I think you could be more precise in documenting the behaviour here.
e.g
"given multiple async failing binds, if invoked asyncronously, then the first invoked will always be the first error returned"

Reply:
It seems that there's no guarantee for order. I ran the original test multiple times and had both successful and failed runs.

Reply:
thanks for checking. That's very interesting! So I guess that means on JS some order to execute the promises is set at runtime, and regardless of the delays, the single threaded execution just means thats the order they are completed in. 🤷‍♂️

Reply:
this made me realise, @michaelbull we should probably be verifying even for the simpler `BindingTest.kt` which binded functions will actually execute in the fail-to-bind test cases - i.e. it would be good to know if side effecting code in a binded function actually ran or not, and that the tests would verify the behaviour.

So in the case of `BindingTest.kt`, we know that no code after the failing bind will execute. So we should verify that. Likewise in `SuspendableBindingTest.kt`.
However in the case of `AsyncSuspendableBindingTest.kt` on jvm, I'm trying to recall whether all the async calls would still run to completion, or whether the cancelaltionException means the running coroutine is just dropped outright. Either way would be nice to have this confirmed by the test. 
Same applies to using async binding JS; we should verify if we expect all the binding code to finish or not, or whether its nondeterministic.

Reply:
referenced in https://github.com/michaelbull/kotlin-result/pull/37

Reply:
@DerYeger given the latest updates in #37 , could you add a test to this class to verify that on a bind failing, the other functions are cancelled/stop executing? See the state change test i added on the other pr for reference

Reply:
@Munzey Sure! That will cause failing checks until the other PR is merged.

On another note, I retested the execution order with higher delays (10ms, 20ms and 1ms) and got different results. It seems that the the small delays of the original test (1ms, 2ms and 1ms) caused issues with scheduling in Kotlin/JS.

Reply:
very nice!
Ha nice that you spotted that. I actually found the same thing could happen on the first recompile of my tests on the other pr when using just a single ms in the difference - I also bumped the ms delay because of that. Not sure if there is an ideal way to test these types of discrepancies. Really just a byproduct of working with asynchronous code I guess.